src/oss-characteristics/CharacteristicTool.cs

```diff
@@ -100,33 +103,33 @@ public CharacteristicTool()
         /// </summary>
         /// <param name="purl">The package-url of the package to analyze.</param>
         /// <returns>List of tags identified</returns>
-        public async Task<Dictionary<string, AnalyzeResult>> AnalyzePackage(PackageURL purl)
+        public async Task<Dictionary<string, AnalyzeResult>> AnalyzePackage(PackageURL purl, string targetDirectoryName)
         {
             Logger.Trace("AnalyzePackage({0})", purl.ToString());
             
             var analysisResults = new Dictionary<string, AnalyzeResult>();
+            List<string> directoryNames = new List<string>();
 
-            string targetDirectoryName = null;
-            while (targetDirectoryName == null || Directory.Exists(targetDirectoryName))
+            var downloadTool = new DownloadTool();
+            // ensure that the cache directory has the required package, download it otherwise
+            bool cached = !string.IsNullOrEmpty(targetDirectoryName);
+            if (cached)
             {
-                targetDirectoryName = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+                directoryNames.AddRange(await downloadTool.EnsureDownloadExists(purl, targetDirectoryName));
```

Comment:
I don't understand this.  It looks like you went from always adding the result of download tool to directoryNames to only doing it if you are cached.  But if you're not cached directoryNames would always be an empty list.

Reply:
That is right, there is a bug...I made multiple iterations and missed the loop. I've abstracted all the package downloading capabilities in a new class PackageDownloader. This should be fixed now, thanks